src/basic/path-util.c

```diff
@@ -599,14 +600,24 @@ static int check_x_access(const char *path, int *ret_fd) {
                 r = access_fd(fd, X_OK);
                 if (r < 0)
                         return r;
-
-                *ret_fd = TAKE_FD(fd);
         } else {
                 /* Let's optimize things a bit by not opening the file if we don't need the fd. */
                 if (access(path, X_OK) < 0)
                         return -errno;
         }
 
+        with_dash = strjoina(path, "/");
+
+        /* If this passes, it must be a directory. */
+        if (access(with_dash, X_OK) >= 0)
+                return -EISDIR;
```

Comment:
hmm, if we have the fd open anyway I think a better check would be using the fd, i.e. do fstat() here